Fix FreeBSD native stdenv again

rhelmot opened the pull request:
Hey gamers it's ya gal Audrey here with another episode of "we need some CI on this derivation". This episode's sponsor is the nix-community builders, who generously offered to set me up with a machine to make sure this build stays building!

Most of these changes are fallout from splitting freebsd.libc into freebsd.libcMinimal + several other derivations. The rest are simply the march of time, rot and let rot.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
  - [x] x86_64-freebsd
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

rhelmot committed:
auto-patchelf: force --keep-libc for BSD hosts

rhelmot committed:
freebsd.libcMinimal: include i18n path; set relevant environment variable in stdenv

rhelmot committed:
freebsd stdenv: switch to llvm bintools

rhelmot committed:
llvmPackages.tblgen: Fix native FreeBSD build

artemist reviewed (changes requested):
Most of the changes seem reasonable, but keep-libc is a weird thing to add with no explanation

rhelmot commented on pkgs/build-support/setup-hooks/auto-patchelf.sh:
```diff
@@ -59,6 +59,13 @@ autoPatchelf() {
+    # shellcheck disable=SC2195
+    case "@hostPlatform@" in
+        *freebsd*|*openbsd*)
+            autoPatchelfFlagsArray+=( --keep-libc )
```
The libc rpath can be stripped out on Linux because ld.so is hardcoded to check within its derivation for libraries, but on platforms with libcMinimal as an input to libc, this won't work because ld.so is in a different derivation than libc and friends. I'll add this as an inline comment.

artemist replied:
Oh, that makes sense, I had a brain fart and thought libc was referring to the dynamic linker.

This sounds like a change that could be made to auto-patchelf, though I'm not confident that's the right place.

rhelmot replied:
Where if not auto-patchelf? The problem is that patchelf strips the path to libc under the assumption that the linker already knows that path, but it doesn't. Copied from DMs for completeness: it occurs to me now that this is maybe fixable without --keep-libc by copying instead of linking ld.so into the libc path, but idk what the fallout of that would be. it might not even work because the paths could be baked in at build time.

artemist replied:
I was talking about changing the default for `keep_libc` in the [auto-patchelf python script](https://github.com/NixOS/nixpkgs/blob/ea2abe305f961b3628aecc55730e11eee0e39c13/pkgs/by-name/au/auto-patchelf/source/auto-patchelf.py), not [the hook](https://github.com/NixOS/nixpkgs/blob/ea2abe305f961b3628aecc55730e11eee0e39c13/pkgs/build-support/setup-hooks/auto-patchelf.sh). The actual [patchelf program](https://github.com/nixos/patchelf) written in C++ does not know about libc at all, so this change has to happen in either the hook or the python script

artemist replied:
I tend to prefer symlinks to copying files, so I would support `--keep-libc` over copying ld.so

rhelmot replied:
I ended up reverting the hook patch and patching the python script. I do think this is better, thanks for the suggestion.

artemist commented:
Testing on my FreeBSD 14.2 VM, I'm still getting a test failure in diffutils (`FAIL: test-time`) when trying to build stdenv. I should be using a fairly recent version of our patched lix with sandboxing.

artemist committed:
libarchve: disable another atime test

This test relies on atime of files and fails inconsistently on FreeBSD.

artemist committed:
diffutils: disable atime-based test on FreeBSD

This atime-based test fails on some FreeBSD configurations

rhelmot committed:
gdbm: fix freebsd native build

Pulls in a patch from freebsd-ports. This patch cannot be made unconditional
without a rework since it modifies the base behavior.

rhelmot committed:
libuv: add more workarounds for tests on FreeBSD

rhelmot committed:
stdenv/freebsd: isolate bootstrap process more

Copying instead of symlinking prevents leakage of the bootstrap FHS
into the 0->1 bootstrap stage. The symptom this addresses is libssh2
failing because it can't find libssl, since it is failing to use the
ld wrapper.

rhelmot committed:
stdenv/freebsd: copy clang lib folder out of bootstrap archive

Fixes the boot0 build of python

rhelmot committed:
llvm: disable a test that fails in the FreeBSD build jail

rhelmot committed:
gnupatch: disable a filesystem-dependent test

artemist commented on pkgs/stdenv/freebsd/linkBootstrap.sh:
```diff
@@ -6,5 +6,9 @@ for path in $paths; do
+    if [[ -d $src/$path ]]; then
+        ln -s $src/$path $out/$path
+    else
+        cp -RL $src/$path $out/$path
```
Why do you need to copy? If this is required to prevent a dependency on the bootstrap tarball contents I'd prefer to fix it somewhere else

rhelmot replied:
Done. I put it in the nix code instead of in the shell code so it isn't part of the derivation hash.

artemist replied:
Thanks, I'll do some more build tests then merge

artemist reviewed (commented):
There's only one thing I'm confused about. I'm building now, if that works I think it should be good to merge.

alyssais commented:
This patch is actually upstream, so could probably have been applied unconditionally relatively safely.
